### PR TITLE
feat(jaegermcp): Add get_trace_stats tool

### LINKED ISSUE #8577: [Feature]: feat(jaegermcp): Add get_trace_stats tool for aggregate trace statistics
### Requirement

As a Jaeger MCP server user, I want to get aggregate statistics across traces matching a filter so that I can understand the overall health and performance of a service without fetching individual traces one by one.

### Problem

The existing search_traces tool returns individual trace summaries. To understand overall error rate or latency percentiles, an LLM agent must fetch and process many individual traces, consuming large amounts of context window.

### Proposal

Add a get_trace_stats tool that accepts the same filters as search_traces but returns aggregate statistics:
- Total trace count and error count / error rate
- Latency percentiles: p50, p95, p99 (microseconds)
- Span count stats: min, max, mean
- Top services by occurrence (capped at 10)

This fits the progressive disclosure pattern already in the MCP extension — agents call get_trace_stats first for a health overview, then drill into individual traces with search_traces or get_span_details.

### Open questions

_No response_

### PR DESCRIPTION
**Pull request by vipulpandey21**
Adds a new `get_trace_stats` MCP tool that computes aggregate statistics (trace count, error count, error rate, latency percentiles p50/p95/p99, span count statistics, and top services) across traces matching the same filters as `search_traces`.

This allows MCP clients and LLM agents to get a high-level health overview of a service before drilling into individual traces, following the progressive disclosure pattern already established in the Jaeger MCP extension.

Resolves #8577

## Which problem is this PR solving?

* Resolves #8577
* The Jaeger MCP extension currently provides `search_traces` for retrieving individual trace summaries, but does not provide a way to obtain aggregate service-level statistics across matching traces.
* As a result, an MCP client or LLM agent would need to fetch and process many individual traces to compute basic metrics such as error rate or latency percentiles, consuming additional requests and context.

## Description of the changes

* Added `get_trace_stats` input/output type definitions in `internal/types/get_trace_stats.go`
* Added `get_trace_stats` handler in `internal/handlers/get_trace_stats.go` that:

  * reuses the same filters as `search_traces`
  * aggregates matching traces using `AggregateTraces`
  * computes trace count, error count, and error rate
  * computes duration statistics (`min`, `max`, `mean`, `p50`, `p95`, `p99`)
  * computes span count statistics
  * returns the top services by occurrence (capped at 10)
* Registered the new tool in `server.go`
* Updated MCP integration tests for tool discovery
* Added 19 unit tests covering:

  * normal execution paths
  * empty results
  * all-error traces
  * percentile calculations
  * service ordering and limits
  * invalid inputs
  * query errors
  * helper function edge cases
* Fixed two pre-existing lint blockers encountered while running `make lint`:

  * handled the returned error from `listener.Close()`
  * renamed `durationMin` / `durationMax` to `minDuration` / `maxDuration` to satisfy `revive`
* Updated the Jaeger MCP README documentation

## How was this change tested?

* Added unit tests for the new handler and helper functions
* Ran `go test` for all `jaegermcp` packages successfully
* Ran lint and formatting checks successfully
* Verified tool registration through existing MCP integration tests

## Checklist

* [x] I have read https://github.com/jaegertracing/jaeger/blob/main/CONTRIBUTING_GUIDELINES.md
* [x] I have signed all commits
* [x] I have added unit tests for the new functionality
* [x] I have run lint and test steps successfully: `make lint test`

## AI Usage in this PR (choose one)

See https://github.com/jaegertracing/jaeger/blob/main/CONTRIBUTING_GUIDELINES.md#ai-usage-policy

* [ ] **None**: No AI tools were used in creating this PR
* [ ] **Light**: AI provided minor assistance (formatting, simple suggestions)
* [x] **Moderate**: AI helped with code generation or debugging specific parts
* [ ] **Heavy**: AI generated most or all of the code changes

### CODE REVIEW

File: cmd/jaeger/internal/extension/jaegermcp/internal/types/get_trace_stats.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) 2026 The Jaeger Authors.
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/types/get_trace_stats.go]: Fixed — removed the UTF-8 BOM and saved the file as plain UTF-8 without BOM.

File: cmd/jaeger/internal/extension/jaegermcp/internal/types/get_trace_stats.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+// DurationStats contains latency percentile statistics across all analysed traces (in microseconds).
+DurationStats DurationStats `json:"duration_stats" jsonschema:"Latency statistics across all analysed traces"`
+
+// SpanStats contains statistics about span counts across all analysed traces.
+SpanStats SpanStats `json:"span_stats" jsonschema:"Span count statistics across all analysed traces"`
+
+// TopServices lists the services that appear most frequently across the analysed traces,
+// ordered by descending occurrence count.
+TopServices []ServiceCount `json:"top_services" jsonschema:"Services ordered by descending occurrence count across traces"`
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/types/get_trace_stats.go]: Fixed along with the BOM removal in the same commit.

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/test_helpers_test.go [hunk 3]
```diff
@@ -17,6 +18,22 @@ import (
 // testTraceID is a common trace ID used across tests
 const testTraceID = "12345678901234567890123456789012"
 
+// uniqueTraceIDs returns n distinct 32-hex-char trace ID strings.
+// Used by FindTraces-based tests where each trace must have a unique ID so
+// that AggregateTraces does not merge them together.
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/test_helpers_test.go]: Fixed — updated the comment to accurately state that AggregateTraces merges spans sharing the same trace ID regardless of how they are yielded, so unique IDs are what prevents merging.

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/test_helpers_test.go [hunk 4]
```diff
@@ -94,12 +111,18 @@ func newMockYieldingEmpty() *mockQueryService {
 	}
 }
 
-// newMockFindTraces creates a mock for FindTraces calls that yields the given traces
+// newMockFindTraces creates a mock for FindTraces calls that yields each trace
+// individually (one per iteration step) so that AggregateTraces treats them as
+// separate traces even when they share the same trace ID.
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/test_helpers_test.go]: Fixed — updated the comment to clarify that yielding individually controls emission order but does not prevent AggregateTraces from merging same-ID traces. Tests requiring distinct aggregated traces should use unique trace IDs.

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_trace_stats.go [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+		result = append(result, types.ServiceCount{
+			Service:    svc,
+			TraceCount: count,
+		})
+	}
+
+	slices.SortStableFunc(result, func(a, b types.ServiceCount) int {
+		if b.TraceCount != a.TraceCount {
+			return b.TraceCount - a.TraceCount
+		}
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_trace_stats.go]: Fixed — replaced b.TraceCount - a.TraceCount with explicit </> comparisons to avoid integer overflow.

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_trace_stats.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+		SpanName:     input.SpanName,
+		Attributes:   input.Attributes,
+		WithErrors:   input.WithErrors,
+		DurationMin:  input.DurationMin,
+		DurationMax:  input.DurationMax,
+		SearchDepth:  input.SearchDepth,
+	}
+	tmp := &searchTracesHandler{maxResults: maxResults}
+	return tmp.buildQuery(searchInput)
+}
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_trace_stats.go:128]: Fixed — extracted the query-building logic into a standalone buildTraceQueryParams function in search_traces.go. Both searchTracesHandler and getTraceStatsHandler now call it directly. The partial struct construction is gone.

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/search_traces.go [hunk 7]
```diff
@@ -152,8 +159,8 @@ func (h *searchTracesHandler) buildQuery(input types.SearchTracesInput) (querysv
 	if searchDepth <= 0 {
 		searchDepth = defaultSearchDepth
 	}
-	if searchDepth > h.maxResults {
-		searchDepth = h.maxResults
+	if searchDepth > maxResults {
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/search_traces.go]: Fixed — added maxResults > 0 guard before the clamp: if maxResults > 0 && searchDepth > maxResults. When maxResults is 0 (no limit), searchDepth is no longer clamped to 0.

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_trace_stats.go [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+		if summary.HasErrors {
+			errorCount++
+		}
+
+		for _, svc := range summary.Services {
+			serviceCounts[svc]++
+		}
+
+		if h.maxResults > 0 && len(durations) >= h.maxResults {
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_trace_stats.go]: Fixed — introduced effectiveLimit = query.TraceQueryParams.SearchDepth (the already-clamped value from buildTraceQueryParams) and use that as the loop break condition instead of h.maxResults.

File: cmd/jaeger/internal/extension/jaegermcp/internal/types/get_trace_stats.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+// DurationMin is the minimum trace duration filter (optional, e.g., "2s", "100ms").
+DurationMin string `json:"duration_min,omitempty" jsonschema:"Minimum duration filter (e.g. 2s 100ms)"`
+
+// DurationMax is the maximum trace duration filter (optional).
+DurationMax string `json:"duration_max,omitempty" jsonschema:"Maximum duration filter (e.g. 10s 1m)"`
+
+// SearchDepth defines the maximum number of traces to analyse for statistics.
+// Default: 100, maximum is controlled by server configuration (MaxSearchResults).
+SearchDepth int `json:"search_depth,omitempty" jsonschema:"Maximum number of traces to analyse (default: 100, max controlled by server config)"`
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/types/get_trace_stats.go]: Fixed — updated the comment and jsonschema tag to say default: 10, matching the actual default in buildTraceQueryParams.

File: cmd/jaeger/internal/extension/jaegermcp/internal/types/get_trace_stats.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+// DurationStats contains latency percentile statistics across all analysed traces (in microseconds).
+DurationStats DurationStats `json:"duration_stats" jsonschema:"Latency statistics across all analysed traces"`
+
+// SpanStats contains statistics about span counts across all analysed traces.
+SpanStats SpanStats `json:"span_stats" jsonschema:"Span count statistics across all analysed traces"`
+
+// TopServices lists the services that appear most frequently across the analysed traces,
+// ordered by descending occurrence count.
+TopServices []ServiceCount `json:"top_services" jsonschema:"Services ordered by descending occurrence count across traces"`
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/types/get_trace_stats.go]: Fixed — removed the UTF-8 BOM and ran gofmt on the file.

File: cmd/jaeger/internal/extension/jaegermcp/server.go [hunk 11]
```diff
@@ -197,4 +199,13 @@ func (s *server) registerTools() {
...
 	}, handlers.NewGetDependenciesHandler(s.queryAPI))
+
+	mcp.AddTool(s.mcpServer, &mcp.Tool{
+		Name: "get_trace_stats",
+		Description: "Compute aggregate statistics (count, error rate, latency percentiles p50/p95/p99, " +
+			"span counts, top services) across traces matching the given filters. " +
+			"Accepts the same filters as search_traces. " +
+			"Use this tool to understand the overall health and performance of a service " +
+			"before drilling into individual traces.",
+	}, handlers.NewGetTraceStatsHandler(s.queryAPI, s.config.MaxSearchResults))
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/server.go:210]: Fixed — updated the description to say "over up to search_depth traces matching the given filters" to make the sampling behavior explicit.

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_trace_stats.go [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+
+	// Use the effective search depth computed by buildStatsQuery (already clamped
+	// to maxResults and defaulted) as the loop limit for deterministic behaviour.
+	effectiveLimit := query.TraceQueryParams.SearchDepth
+
+	tracesIter := h.queryService.FindTraces(ctx, query)
+	aggregatedIter := jptrace.AggregateTraces(tracesIter)
+
+	var durations []int64
+	var spanCounts []int
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_trace_stats.go:64]: Fixed — added pre-allocation with capacity effectiveLimit when effectiveLimit > 0.

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/test_helpers_test.go [hunk 13]
```diff
@@ -17,6 +18,23 @@ import (
...
+// the string, not decoded hex bytes).
+func uniqueTraceIDs(n int) []string {
+	ids := make([]string, n)
+	for i := range ids {
+		// Put the index in the first 16 characters (as big-endian hex) so the
+		// 16-byte string copy captures the difference.
+		ids[i] = fmt.Sprintf("%016x%016x", i+1, i+1)
+	}
+	return ids
+}
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/test_helpers_test.go]: Fixed — updated uniqueTraceIDs to generate proper 32-char hex strings (fmt.Sprintf("%032x", i+1)) and updated createTestTrace to hex-decode the string into pcommon.TraceID bytes using encoding/hex. Added a fallback for legacy callers passing non-hex strings like testTraceID.

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/test_helpers_test.go [hunk 14]
```diff
@@ -94,18 +110,26 @@ func newMockYieldingEmpty() *mockQueryService {
...
-				yield(traces, nil)
+				for _, t := range traces {
+					if !yield([]ptrace.Traces{t}, nil) {
+						return
+					}
+				}
 			}
 		},
 	}
 }
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/test_helpers_test.go:129]: The actual code already yields one trace at a time (the for loop with individual yield calls). The diff Copilot saw showing a bare yield(traces, nil) line was a display artifact from the diff context — that line does not exist in the file. No change needed here.

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/test_helpers_test.go [hunk 15]
```diff
@@ -116,9 +140,15 @@ func createTestTrace(traceID string, serviceName string, spanName string, hasErr
...
-	// Set trace ID
-	tid := pcommon.TraceID{}
-	copy(tid[:], traceID)
+	// Hex-decode the trace ID string into a pcommon.TraceID so the resulting
+	// TraceID bytes match what a real trace would produce.
+	var tid pcommon.TraceID
+	if b, err := hex.DecodeString(traceID); err == nil && len(b) == 16 {
+		copy(tid[:], b)
+	} else {
+		// Fallback for legacy callers passing non-hex strings (e.g. testTraceID).
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/test_helpers_test.go]: Fixed — updated the comment to say "Fallback for callers passing non-hex strings or values with the wrong decoded length."

File: cmd/jaeger/internal/extension/jaegermcp/internal/types/get_trace_stats.go [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+	WithErrors bool `json:"with_errors,omitempty" jsonschema:"If true only analyse traces containing error spans"`
+
+	// DurationMin is the minimum trace duration filter (optional, e.g., "2s", "100ms").
+	DurationMin string `json:"duration_min,omitempty" jsonschema:"Minimum duration filter (e.g. 2s 100ms)"`
+
+	// DurationMax is the maximum trace duration filter (optional).
+	DurationMax string `json:"duration_max,omitempty" jsonschema:"Maximum duration filter (e.g. 10s 1m)"`
+
+	// SearchDepth defines the maximum number of traces to analyse for statistics.
+	// Default: 10, maximum is controlled by server configuration (MaxSearchResults).
```
[the remark this answers is not in the document]
Comment (vipulpandey21) [reply cmd/jaeger/internal/extension/jaegermcp/internal/types/get_trace_stats.go]: Fixed — standardized to American English (analyze/analyzed) throughout the types and handler files to match the rest of the repo.
